ansible/roles/test/files/ptftests/vxlan_traffic.py

```diff
@@ -516,3 +517,258 @@ def test_encap(
 
         finally:
             Logger.info("")
+
+
+class VxLAN_in_VxLAN(VXLAN):
+    def test_encap(
+            self,
+            ptf_port,
+            vni,
+            ptf_addr,
+            destination,
+            nhs,
+            test_ecn=False,
+            random_dport=True,
+            random_sport=False,
+            random_src_ip=False):
+        '''
+           Test the encapsulation of packets works correctly when the payload
+           itself is a vxlan packet.
+           1. Send a TCP packet to the DUT port.
+           2. Verify that the DUT returns an encapsulated packet correctly.
+           3. Optionally: Perform if the ECMP is working(all nexthops are used
+           equally).
+        '''
+        pkt_len = 100
+        pkt_opts = {
+            "pktlen": pkt_len,
+            "eth_dst": "aa:bb:cc:dd:ee:ff",
+            "eth_src": "ff:ee:dd:cc:bb:aa",
+            "ip_dst": "1.1.1.1",
+            "ip_src": "2.2.2.2",
+            "ip_id": 105,
+            "ip_ttl": 64,
+            "tcp_sport": 3000,
+            "tcp_dport": 5000}
+        innermost_frame = simple_tcp_packet(**pkt_opts)
+
+        try:
+            pkt_len = 100
+            udp_dport = self.vxlan_port
+
+            options = {'ip_ecn': 0}
+            options_v6 = {'ipv6_ecn': 0}
+            if test_ecn:
+                ecn = random.randint(0, 3)
+                options = {'ip_ecn': ecn}
+                options_v6 = {'ipv6_ecn': ecn}
+
+            # ECMP support, assume it is a string of comma seperated list of
+            # addresses.
+            check_ecmp = False
+            working_nhs = list(set(nhs) - set(self.downed_endpoints))
+            expect_success = self.expect_encap_success
+            test_nhs = working_nhs
+            packet_count = self.packet_count
+            if not working_nhs:
+                # Since there is no NH that is up for this destination,
+                # we can't expect success here.
+                expect_success = False
+                test_nhs = nhs
+                # Also reduce the packet count, since this script has to wait
+                # 1 second per packet(1000 packets is 20 minutes).
+                packet_count = 4
+            returned_ip_addresses = {}
+            for host_address in test_nhs:
+                check_ecmp = True
+                # This will ensure that every nh is used atleast once.
+                Logger.info(
+                    "Sending %s packets from port %s to %s",
+                    packet_count,
+                    str(ptf_port),
+                    destination)
+                for _ in range(packet_count):
+                    udp_sport = get_incremental_value('udp_sport')
+                    if isinstance(ip_address(destination), IPv4Address) and \
+                            isinstance(ip_address(ptf_addr), IPv4Address):
+                        if random_src_ip:
+                            ptf_addr = get_ip_address(
```

Comment:
The `get_ip_address` function is not defined and imported anywhere. Wondering how this would work?

Reply:
@wangxin : It should be seen at line 79 in the same script.